### ISSUE TITLE
[Bug]: Using typescript: "react-docgen-typescript" breaks source preview

### ISSUE THREAD

**Issue by adrianaferrugento** [storybookjs/storybook#26510]
### Describe the bug

Using the following in my main config breaks the preview source code.

```
  typescript: {
    reactDocgen: 'react-docgen-typescript',
  },
```
![Screenshot 2024-03-15 at 09 53 15](https://github.com/storybookjs/storybook/assets/76945696/f822a8b9-0223-47e3-8452-eaaf182b810d)
![Screenshot 2024-03-15 at 09 53 07](https://github.com/storybookjs/storybook/assets/76945696/de2422b5-411f-41e9-8241-cd087e8fccdc)

### To Reproduce

https://stackblitz.com/edit/github-cuvfy3?file=.storybook%2Fmain.ts

### System

```bash
My local system info


Storybook Environment Info:

  System:
    OS: macOS 14.4
    CPU: (10) arm64 Apple M1 Pro
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.11.1 - ~/.nvm/versions/node/v20.11.1/bin/node
    npm: 10.2.4 - ~/.nvm/versions/node/v20.11.1/bin/npm
    pnpm: 8.15.4 - ~/.nvm/versions/node/v20.11.1/bin/pnpm <----- active
  Browsers:
    Chrome: 122.0.6261.129
    Safari: 17.4
  npmPackages:
    @storybook/addon-a11y: 8.0.0 => 8.0.0
    @storybook/addon-docs: 8.0.0 => 8.0.0
    @storybook/addon-essentials: 8.0.0 => 8.0.0
    @storybook/addon-interactions: 8.0.0 => 8.0.0
    @storybook/addon-mdx-gfm: 8.0.0 => 8.0.0
    @storybook/addon-themes: 8.0.0 => 8.0.0
    @storybook/addon-viewport: 8.0.0 => 8.0.0
    @storybook/blocks: 8.0.0 => 8.0.0
    @storybook/core-server: 8.0.0 => 8.0.0
    @storybook/manager-api: 8.0.0 => 8.0.0
    @storybook/preview-api: 8.0.0 => 8.0.0
    @storybook/react: 8.0.0 => 8.0.0
    @storybook/react-vite: 8.0.0 => 8.0.0
    @storybook/react-webpack5: 8.0.0 => 8.0.0
    @storybook/testing-library: 0.2.2 => 0.2.2
    @storybook/theming: 8.0.0 => 8.0.0
    @storybook/types: 8.0.0 => 8.0.0
    chromatic: 11.0.8 => 11.0.8
    eslint-plugin-storybook: 0.8.0 => 0.8.0
    storybook: 8.0.0 => 8.0.0
    storybook-mock-date-decorator: 1.0.1 => 1.0.1
```
```


### Additional context

_No response_

**Comment by vanessayuenn** [comment]
@adrianaferrugento looks like this issue can be mitigated by using `react-docgen` instead, right?

**Comment by ingvaldlorentzen** [comment]
> @adrianaferrugento looks like this issue can be mitigated by using `react-docgen` instead, right?

It can, but for me I loose a lot of type details when using`react-docgen` 
E.g. 
`react-docgen`
<img width="1012" alt="image" src="https://github.com/storybookjs/storybook/assets/8466369/6eb7dddf-ef89-4a1e-bed2-7f085186fd37">
`react-docgen-typescript`
<img width="1029" alt="image" src="https://github.com/storybookjs/storybook/assets/8466369/c57719eb-f059-41d8-b0df-b452bf4f928a">
(this is a Font Awesome `SizeProp`, just used as an example)

So now I have to choose between better types in the controls, or nice code in the examples, which is a bit of a bummer.

**Comment by adrianaferrugento** [comment]
> @adrianaferrugento looks like this issue can be mitigated by using `react-docgen` instead, right?

I actually get some errors when I try to use it with my configuration for some reason. I don't think it's a storybook issue because it works in the reproduction link but if anyone has any ideas let me know. Also losing the type details would be a bit annoying so my preferred solution would be to use `react-docgen-typescript`

![Screenshot 2024-03-15 at 14 59 32](https://github.com/storybookjs/storybook/assets/76945696/81825fe9-9437-4ac8-8c95-a2e23f01eb18)

**Comment by Vempejul** [comment]
I am seeing the same issues, just updated from `7.5.2` to `8.0.0` and had to use `react-docgen-typescript` because I noticed that the automatic `argTypes` generation for `Controls` was missing. Unfortunately in doing so the source preview of my components did not show correctly.

```ts
/* .storybook/main.ts */
import { dirname, join } from "path";
import type { StorybookConfig } from "@storybook/react-vite";
const config: StorybookConfig = {
  framework: {
    name: getAbsolutePath("@storybook/react-vite"),
    options: {},
  },

  stories: ["../stories/**/*.mdx", "../stories/**/*.stories.@(js|jsx|ts|tsx)"],

  addons: [
    getAbsolutePath("@storybook/addon-links"),
    getAbsolutePath("@storybook/addon-essentials"),
    getAbsolutePath("@storybook/addon-interactions"),
  ],

  docs: {
    autodocs: "tag",
  },

  typescript: {
    reactDocgen: "react-docgen-typescript",
    reactDocgenTypescriptOptions: {
      shouldExtractLiteralValuesFromEnum: true,
      shouldExtractValuesFromUnion: true,
      shouldRemoveUndefinedFromOptional: true,
      propFilter: (prop) => {
        if (prop.declarations !== undefined && prop.declarations.length > 0) {
          const hasPropAdditionalDescription = prop.declarations.find(
            (declaration) => {
              return !declaration.fileName.includes("node_modules");
            }
          );

          return Boolean(hasPropAdditionalDescription);
        }

        return true;
      },
    },
  },

  async viteFinal(config) {
    const { mergeConfig } = await import("vite");

    return mergeConfig(config, {});
  },
};

export default config;

function getAbsolutePath(value: string): any {
  return dirname(require.resolve(join(value, "package.json")));
}
```

```tsx
/* Button/index.tsx */
import styles from "./index.module.css";
import { cva, type VariantProps } from "class-variance-authority";

const buttonCVA = cva(styles.base, {
  variants: {
    variant: {
      primary: styles.primary,
      secondary: styles.secondary,
      tertiary: styles.tertiary,
      ghost: styles.ghost,
    },
    buttonTheme: {
      black: styles.themeBlack,
      blue: styles.themeBlue,
      green: styles.themeGreen,
      lilac: styles.themeLilac,
      rose: styles.themeRose,
    },
    icon: {
      left: styles.iconLeft,
      right: styles.iconRight,
      only: styles.iconOnly,
    },
    slim: {
      true: styles.slim,
    },
    squared: {
      true: styles.squared,
    },
    isLoading: {
      true: styles.isLoading,
    },
  },
  defaultVariants: {
    variant: "primary",
    buttonTheme: "green",
  },
});

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonCVA> {}

export const Button = ({
  variant,
  buttonTheme,
  icon,
  slim,
  squared,
  isLoading,
  className,
  children,
  ...props
}: ButtonProps) => {
  return (
    <button
      className={buttonCVA({
        variant,
        buttonTheme,
        icon,
        slim,
        squared,
        isLoading,
        className,
      })}
      {...props}
    >
      {isLoading ? (
        <div className={styles.animationContainer}>
          <svg viewBox={`0 0 32 32`} className={styles.spinnerContainer}>
            <circle r={16} cx={16} cy={16} className={styles.spinner} />
          </svg>
        </div>
      ) : (
        children
      )}
    </button>
  );
```

```tsx
/* button.stories.tsx */
import { Button, type ButtonProps } from "@mypackage/button/src";
import { BasketIcon } from "@mypackage/icon/src";

export default {
  component: Button,
};

export const Component = {
  render: ({ children, ...args }: ButtonProps) => (
    <Button {...args}>
      {args?.icon === "only" ? (
        <BasketIcon />
      ) : (
        <>
          {args?.icon === "left" && (
            <BasketIcon style={{ marginRight: "8px" }} />
          )}
          {children}
          {args?.icon === "right" && (
            <BasketIcon style={{ marginLeft: "8px" }} />
          )}
        </>
      )}
    </Button>
  ),
};
```

Not using `react-docgen-typescript` (before upgrade to `8.0.0`) 
![image](https://github.com/storybookjs/storybook/assets/34095867/80ad7e63-2e0d-40d4-96b0-d0bcdbcbb200)

Not using `react-docgen-typescript` no `argTypes` (Controls) generated (after upgrade to `8.0.0`) 
![image](https://github.com/storybookjs/storybook/assets/34095867/de3c24f7-f03b-4a45-9ce6-fc79973dcbca)

Using `react-docgen-typescript`, `argTypes` (Controls) generated but code preview broken (after upgrade to `8.0.0`) 
![image](https://github.com/storybookjs/storybook/assets/34095867/80bf9723-0a0a-47c3-963c-3ae699244794)
![image](https://github.com/storybookjs/storybook/assets/34095867/c462175e-dd7d-4620-91fe-8473ec98fda6)

**Comment by unleashit** [comment]
Same issue and started cropping up around the time of RC5. Using either react-docgen or react-docgen-typescript. Are you guys in a monorepo? I tried to reproduce outside of my project in a clean install and couldn't.

**Comment by unleashit** [comment]
Pretty much narrowed it down to the 8.0.0-rc.4 package of @storybook/react. If you downgrade it to rc4 for that package, clear all caches (browser + delete node_modules and run `nx reset` if using nx, etc) and then npm install it works again.

Maybe this PR: https://github.com/storybookjs/storybook/pull/26382 ?

**Comment by Vempejul** [comment]
Tried with `8.0.0-rc.3` instead of `8.0.0` and that did the trick.

**Comment by asyrul21** [comment]
Opened duplicated [26538](https://github.com/storybookjs/storybook/discussions/26538) to report same issue. Had to downgrade both `@storybook/react` and `@storybook/react-webpack5` to `8.0.0-rc.3` to fix the broken code preview, though it would be great to have it fixed on 8.0.0 as well. Also, the downgrade would cause some compatibility warnings to appear on the logs.

<img width="926" alt="Screenshot 2024-03-18 at 10 43 07 AM" src="https://github.com/storybookjs/storybook/assets/29390954/8a0618b8-8662-4a15-aa81-1c630ffb4ddf">

Much thanks to all involved with the fix !

**Comment by filiptammergard** [comment]
I'm also getting `<object Object>` instead of `<Component>` when upgrading to v8 using `react-docgen-typescript`.

<img width="405" alt="image" src="https://github.com/storybookjs/storybook/assets/44197016/6688ec84-20f7-4cac-82c2-cc77c1ee7770">

**Comment by kasperpeulen** [comment]
Will be fixed with https://github.com/storybookjs/storybook/pull/26566

**Comment by yannbf** [comment]
Hey everyone! This issue is solved in #26566 and will be released soon. Thanks for your patience 🙏

**Comment by unleashit** [comment]
Thanks @yannbf. I know this deserves a new issue (and I will later when I have time), but the fix resulted in problems with the control display. See the screenshot. React element objects are displayed instead of something like a set object button:

![image](https://github.com/storybookjs/storybook/assets/5051359/54c6eeb0-567e-4b83-8cb4-db9af5f5b67d)

This is using the default react-docgen.

**Comment by yannbf** [comment]
> Thanks @yannbf. I know this deserves a new issue (and I will later when I have time), but the fix resulted in problems with the control display. See the screenshot. React element objects are displayed instead of something like a set object button:
> 
> ![image](https://private-user-images.githubusercontent.com/5051359/314263308-54c6eeb0-567e-4b83-8cb4-db9af5f5b67d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE0NDkxMjYsIm5iZiI6MTcxMTQ0ODgyNiwicGF0aCI6Ii81MDUxMzU5LzMxNDI2MzMwOC01NGM2ZWViMC01NjdlLTRiODMtOGNiNC1kYjlhZjVmNWI2N2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMjZUMTAyNzA2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjhjM2YwNTY1MDlmYmQyZDNiMWFkYjI2ZjAzNTUzMzFhM2ZjODJiMTA1MWQzZmU1NjMyNDFhNzI1NDFlYTMxNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.NdW5F7qo973-2Kb6nPh1fiP5aEXnrRtS-5c5dcfnMos)
> 
> This is using the default react-docgen.

Hey there, I think this has always been the case, as a current limitation of how controls are showing JSX elements. Can you show a reproduction where it behaved differently?

**Comment by unleashit** [comment]
Hey thanks for the reply. So I just figured out that using the LinkTo from the @storybook/addon-links/react in an arg (which allows me to link between stories using the Storybook router) causes this issue. This confused me because any other JSX/ComponentType args would display a set object button.

Honestly, I am having a lot of issues in general with the autodocs. I would normally be happy to take the time to file bugs and even do a repro, but I've spent so much time on it that I've had to switch gears. To be fair I don't doubt it's tough on your end when you don't have control of the main dependency (react-docgen).

With how things are right now in this project, I'll either have to either have an argType for most/all args with various overrides or it might just be easier to create mdx pages with manual prop tables. But I'm still very appreciative about Storybook in general. Thanks for all you do!
